### PR TITLE
Add BMH live-iso proposal

### PR DESCRIPTION
**Pull request by hardys**
Adds a proposal to expose an Ironic API that allows us to boot an arbitrary live-iso image, vs only specifying an image to be written to disk by IPA.

### CODE REVIEW

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dhellmann) [inline design/baremetal-operator/bmh_boot_iso.md:50]: The `image` struct has several other fields related to each other. Would we use those? For example, do we need to verify the checksum of the ISO?

Could we trigger this different behavior if the `status.image.diskFormat` field is set to `ISO`? Or do we already support writing `ISO` images to disk?

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dhellmann) [inline design/baremetal-operator/bmh_boot_iso.md:67]: Does that feature use different states in the Ironic state machine? Do we need corresponding changes in the state machine here in metal3, or will all of the changes be within the Ironic provisioner?

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dhellmann) [inline design/baremetal-operator/bmh_boot_iso.md:92]: We have had other people ask for a way to disable cleaning on deprovisioning. I think we can consider dealing with that via a separate design proposal.

Comment (maelk) [reply design/baremetal-operator/bmh_boot_iso.md:92]: @fmuyassarov is working on a proposal for that already

Comment (fmuyassarov) [reply design/baremetal-operator/bmh_boot_iso.md:92]: Yes, I will open a proposal in metal3-docs for having a configurable disk cleaning, latest end of this week.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:50]: Currently the boot_iso interface Ironic exposes does not require a checksum and I've not checked if the redfish attachment API supports that @dtantsur any idea?  Given that Ironic doesn't require it and metal3 won't download the image to verify any checksum, I omitted that for now.

I don't believe that Ironic has any support for writing ISO images to disk, so we probably could use `status.image.diskFormat` provided we expect that to remain the case?

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:67]: The state transition is different - as soon as the ISO gets booted, the node is marked ACTIVE, which I think should work without changes to the metal3 state machine?

I did wonder if we should also allow booting ramdisk images via PXE since https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html supports both modes, and the state machine behavior is the same (then we could support this "live image boot" functionality for all BMC protocols, at the expense of potentially another image type/field)

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:92]: I think I saw a comment that we're cleaning on both provisioning (the ironic `provide` state transition) and deprovisioning, is the plan to either remove one or make both configurable?

My concern is that for this use-case, we want to boot the FCOS installer iso as fast as possible, and there is already code in the image to handle wiping disks etc.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:67]: If we support the PXE option, I'm not sure if the `status.image.diskFormat` approach will work well, since we'll need a way to pass both a kernel and ramdisk URL IIUC (that's not explicitly mentioned in the [ironic docs](https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html#configuration) but I assume we'll need to override the default IPA kernel/ramdisk in driver_info)

File: design/baremetal-operator/bmh_boot_iso.md
Comment (fmuyassarov) [reply design/baremetal-operator/bmh_boot_iso.md:92]: The plan is to be able to completely disable/enable disk cleaning from metal3 per node. This requires some changes in the Ironic code as well, because, once we enable disk cleaning for a node, there is no way back to disable it again.

So, currently if we disable automated cleaning globally, then we can enable it only for certain nodes that require disk cleaning as part of deprovisioning. However, from the security perspective the safer option would be to enable automated cleaning globally and then disable it for certain nodes that don't require disk cleaning. And this option is not supported currently. 

We had quick chat on: https://github.com/metal3-io/baremetal-operator/issues/626 and the issue is opened in Ironic https://storyboard.openstack.org/#!/story/2008113.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dhellmann) [reply design/baremetal-operator/bmh_boot_iso.md:50]: OK, what I'm thinking about here is the fact that the fields under `spec.image` are all currently closely related. This new field is sort of related, but it's also sort of standalone because it sounds like the user would either set `spec.image.iso` or all of the other fields (with `diskFormat` remaining optional).

So, if the ISO URL isn't going to use the other image settings, perhaps it should be in its own group. You mention "live boot" in another comment, so how about `spec.liveBoot.url`? That would be parallel to the `spec.image.url` field, and give us a place to put any other settings we might need related to managing the live booting process. For example, if we decide to support PXE booting then we could have `spec.liveBoot.kernelURL` and `spec.liveBoot.ramDiskURL`.

Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:50]: @dhellmann thanks for the feedback - `spec.liveBoot.*` sounds good to me - do you think it'd be reasonable to rework this proposal so we cover both the live-iso and PXE ramdisk cases supported by ironic?

Comment (dhellmann) [reply design/baremetal-operator/bmh_boot_iso.md:50]: I could go either way on the PXE support. If we actually need it or want to support it now, then I could see including it in this proposal. If we don't need it right now, we could mention it as future work to explain why the API change is designed as it is and then have a separate enhancement later when we're ready to actually support it.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:67]: @dhellmann actually @dtantsur mentioned that Ironic can already boot an ISO image via ipxe, so perhaps using diskFormat would be enough - I've not yet tried that though.

Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:67]: Ok to clarify we'd always specify an iso, but internally set a different boot_interface depending on whether it's virtualmedia or not https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html so if we don't currently see a requirement for separate kernel/ramdisk I can adjust this propoal to specify an iso that can be booted via ipxe or virtualmedia transparently :)

Comment (dhellmann) [reply design/baremetal-operator/bmh_boot_iso.md:67]: I've grown to like the separate `liveBoot` approach, because it feels a bit more explicit about expressing the intent for how to treat the image. I think `diskFormat` will work fine, too, though. I don't think I feel strongly enough about either approach to hold up progress over it.

@zaneb do you have an opinion?

Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:67]: I'm fine with `liveBoot` and agree it makes it a little clearer it's a different interface/behavior to the current `image` section.

If we're bikeshedding interfaces I'd perhaps prefer `liveImage` since it aligns slightly better with the existing API but I'm happy to go with the consensus :)

Comment (dhellmann) [reply design/baremetal-operator/bmh_boot_iso.md:67]: I don't know about a checksum but @hardys did mention additional URLs for the ramdisk and kernel. I wouldn't want to add those to `spec.image` because it would be easy to get mixed up about which combination of fields work together and which are ignored in each case. A separate section makes that obvious, even at the expense of possibly having to have "duplicate" fields for some things.

So, how about a `spec.liveImage.url` for now to make it easy to add those other fields later?

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_boot_iso.md:31]: Probably a better link: https://docs.openstack.org/ironic/latest/admin/ramdisk-boot.html

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_boot_iso.md:40]: Ironic supports booting ISOs with iPXE as well, I think we should include it too.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_boot_iso.md:61]: We're planning on supporting user data with redfish-virtual-media ISO boot (but not iPXE).

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [reply design/baremetal-operator/bmh_boot_iso.md:67]: I like `liveImage.url`.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_boot_iso.md:80]: On the roadmap.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_boot_iso.md:86]: Unfortunately, currently fast-track does not really work with virtual media at all (reboots always happen). I'll be working on it in the coming weeks.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [reply design/baremetal-operator/bmh_boot_iso.md:92]: A patch is on review for this feature.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_boot_iso.md:102]: Add the discussed future enhancements as soft dependencies?

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:67]: Ok I'll go with `spec.liveImage.url` as that seems to be the consensus, and it leaves the option of adding more fields in future if needed, thanks!

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:61]: Thanks, is there an upstream ironic link I can include here to reference that work?

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:80]: Do you have a link I can reference please @dtantsur ?

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md:102]: Yes if you can provide links to the Ironic planned work we can include them here, or I'll add links inline with the text?

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md]: Ok I've mentioned that we can boot an iso via iPXE but that this proposal doesn't consider kernel/ramdisk boot, only an iso

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md]: Added a link to https://github.com/metal3-io/metal3-docs/pull/151

**Comment by hardys** [comment]
Ok thanks for the feedback all - I've pushed an update which I think addresses the feedback so far :)

File: design/baremetal-operator/bmh_live_boot.md
Comment (dhellmann) [inline design/baremetal-operator/bmh_live_boot.md:81]: We wouldn't do it as part of this enhancement, but should we add a flag to disable inspection to support workflows that get the hardware details another way?

File: design/baremetal-operator/bmh_live_boot.md
Comment (dhellmann) [inline design/baremetal-operator/bmh_live_boot.md:99]: Is this still a TODO?

**Review by dhellmann — COMMENTED** [review]
I'm happy with the state of this. There are a few details we could fill in as updates to this PR or as separate PRs later.

/approve

**Review by dhellmann — APPROVED** [review]

File: design/baremetal-operator/bmh_live_boot.md
Comment (hardys) [reply design/baremetal-operator/bmh_live_boot.md:81]: Yes we could do, I was thinking we'd rely on the existing interface e.g pass some status annotation at the point of creating the BMH, but then we'd need to change behavior so it can be updated after the hardware details are collected.  

A flag may be clearer though, I can either include that here or in a follow-up proposal when we have the boot-iso part working?

File: design/baremetal-operator/bmh_live_boot.md
Comment (hardys) [reply design/baremetal-operator/bmh_live_boot.md:99]: Yes I've asked for some links from @dtantsur in the comments above, we can merge without if needed though and I'll add them as a follow-up

File: design/baremetal-operator/bmh_live_boot.md
Comment (dhellmann) [reply design/baremetal-operator/bmh_live_boot.md:81]: For the use case you're working on, the host resource needs to exist before the ISO with the alternative agent is booted, so there's no opportunity to collect valid data in advance. We could pass fake data, but that feels wrong.

How about if we say that by default if there is a live image on the host we would not perform inspection, under the assumption that we either don't care about hardware details in those cases or the details will come from somewhere else later. Then we can use other design docs to cover an API for setting hardware details after the first reconciliation and to add a flag like `spec.liveImage.enableInspection` to restore the ability to (optionally) perform inspection with IPA before attaching the ISO.

File: design/baremetal-operator/bmh_live_boot.md
Comment (dhellmann) [reply design/baremetal-operator/bmh_live_boot.md:99]: Yeah, I'm trying to help us keep moving. These documents are meant to be kept up to date, so I think it's OK to add links here later if we know there is work planned but don't have artifacts for that work, yet.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [reply design/baremetal-operator/bmh_boot_iso.md]: Just created https://storyboard.openstack.org/#!/story/2008380

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [reply design/baremetal-operator/bmh_boot_iso.md]: https://storyboard.openstack.org/#!/story/2008363

File: design/baremetal-operator/bmh_live_boot.md
Comment (dtantsur) [reply design/baremetal-operator/bmh_live_boot.md:99]: The links have been provided, see above.

File: design/baremetal-operator/bmh_live_boot.md
Comment (hardys) [reply design/baremetal-operator/bmh_live_boot.md:81]: I'm not sure making inspection conditional on the spec.liveImage will work, because we may create the BMH resource then set the image reference later, and in that case we want the BMH to enroll and be ready to boot the iso, but not perform the existing inspection steps?

Perhaps adding an explicit flag elsewhere in the spec would be clearer?

File: design/baremetal-operator/bmh_live_boot.md
Comment (zaneb) [inline design/baremetal-operator/bmh_live_boot.md:60]: It's about time we added a webhook to validate this kind of stuff.

File: design/baremetal-operator/bmh_live_boot.md
Comment (zaneb) [reply design/baremetal-operator/bmh_live_boot.md:81]: Let's just have an annotation that causes inspection to be skipped. That's simple to do, works for all use cases, and avoids making it part of the Spec which is a bit icky.
There is a proposal I can't find right now to add an annotation to force a new inspection. Maybe we could use the same key with different values (e.g. `never` to skip it even the first time vs. `refresh` to trigger a new one).

Note that to add the data externally later on you'll need RBAC write access to the Status, which is unusual.

**Review by zaneb — COMMENTED** [review]
/approve

File: design/baremetal-operator/bmh_live_boot.md
Comment (hardys) [reply design/baremetal-operator/bmh_live_boot.md:81]: @zaneb I think this deserves a separate enhancement proposal - I suspect we need a way to skip inspection, but also add the data via the existing status annotation (avoiding the status RBAC issue you mention), which beyond the scope of this PR - I'll start a new one if that sounds reasonable.

File: design/baremetal-operator/bmh_live_boot.md
Comment (hardys) [reply design/baremetal-operator/bmh_live_boot.md]: I added a new commit with the links, thanks!

File: design/baremetal-operator/bmh_boot_iso.md
Comment (hardys) [reply design/baremetal-operator/bmh_boot_iso.md]: @dtantsur thanks, if you have a link for this please let me know, and I'll add it to the list of Ironic roadmap items

**Comment by hardys** [comment]
Ok I added the Ironic links, and if folks agree I think the skipping inspection workflow would be best discussed in a new proposal - implementation of this doesn't directly depend on it, so I'll create a new PR where we can decide the best approach for that.

File: design/baremetal-operator/bmh_live_boot.md
Comment (dhellmann) [reply design/baremetal-operator/bmh_live_boot.md:81]: I agree with moving the inspection discussion to another enhancement.

**Review by dhellmann — APPROVED** [review]

**Comment by hardys** [comment]
I think this is now ready to merge?

We'll also need some way to disable cleaning as discussed via https://github.com/metal3-io/metal3-docs/pull/151 and I'll raise a new enhancement to discuss how we can skip inspection.

File: design/baremetal-operator/bmh_live_boot.md
Comment (dhellmann) [reply design/baremetal-operator/bmh_live_boot.md:60]: There's a period of time when validation web hooks aren't applied, between the time the CRD is installed (and the API server supports the new resource type) and when the web hook server is registered and starts.

**Comment by dhellmann** [comment]
> I think this is now ready to merge?

I think so. @zaneb approved an earlier version and I'm happy with this version. It would be good to have a thumbs-up from someone else, too.

Maybe we can start the lazy consensus process this week?

**Comment by hardys** [comment]
Discussion on the implementation PR and slack indicates we may want to go with a different interface so I'll hold this until that direction is agreed.

/hold

File: design/baremetal-operator/bmh_live_boot.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_live_boot.md:77]: There is now! Not sure we have any documentation about it, but ping @bfournie or myself for pointers.

File: design/baremetal-operator/bmh_live_boot.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_live_boot.md:87]: For clarity: in-band cleaning. There is also out-of-band cleaning that is done via talking to the BMC, and we'll keep using that. It is only available as manual cleaning now, so it's fine to disable the automatic one.

File: design/baremetal-operator/bmh_live_boot.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_live_boot.md:105]: Merged

File: design/baremetal-operator/bmh_live_boot.md
Comment (dtantsur) [inline design/baremetal-operator/bmh_live_boot.md:103]: On review, will try to get it merged this week.

File: design/baremetal-operator/bmh_boot_iso.md
Comment (dtantsur) [reply design/baremetal-operator/bmh_boot_iso.md]: https://storyboard.openstack.org/#!/story/2008321

**Comment by hardys** [comment]
Alternative BMH API proposed via https://github.com/metal3-io/baremetal-operator/pull/759 - if we're in agreement with that I'll rework this to reflect the new interface.

**Comment by hardys** [comment]
Updated to reflect interface in https://github.com/metal3-io/baremetal-operator/pull/759

/hold cancel

**Comment by hardys** [comment]
The implementation landed ref https://github.com/metal3-io/baremetal-operator/pull/759 so it'd be good to merge this - @zaneb could you revisit perhaps?  Thanks!

**Comment by dhellmann** [comment]
/retest
/lgtm

Since the code has merged, let's land this and update it in another PR if we're missing any details.

**Comment by zaneb** [comment]
Netlify tests are failing with `make: *** No rule to make target 'build'.  Stop.`, so I would infer that this needs to be rebased to pick up whatever tooling support was added in the repo for those jobs.
(I miss Zuul.)

**Comment by hardys** [comment]
> Netlify tests are failing with `make: *** No rule to make target 'build'. Stop.`, so I would infer that this needs to be rebased to pick up whatever tooling support was added in the repo for those jobs.
> (I miss Zuul.)

Thanks, now rebased and the jobs completed OK

**Pull request merged**
